### PR TITLE
TWILL-21: Multiple Block Editors

### PR DESCRIPTION
**Pull request by ptrckvzn**
## Description

This PR integrates the frontend work of #112, rebased on the `2.x` branch.

It adds a migration to the table `blocks` in order to persist the "editor_name" in the database and retrieve them correctly both in the admin form and on the frontend with the introduction of a new function `renderNamedBlocks('fantastic-name', ...)`.

Multiple blocks editor are defined with the `name` key on the `@formField` statement in the module's form.

```php
// Use 'default' block name (compatible with previous block editor)
@formField('block_editor', [
  'type' => ['text', 'image', 'accordion'],
]);

// Use a named block editor
@formField('block_editor', [
  'name' => 'fantastic-name',
  'label' => 'Fantastic name', // title to display in the Editor dropdown selector
  'trigger' => 'Add fantastic content', // custom trigger label
  'blocks' => ['text', 'image', 'accordion'],
]);
```

On the frontend, you can retrieve and output the blocks with:

```php
// Render 'default' or null blocks
{!! $item->renderBlocks(); !!}

// Render 'fantastic-name' blocks
{!! $item->renderNamedBlocks('fantastic-name'); !!}
```

Add scope to Block model.

```php
$item->blocks()->editor('fantastic-name')->get();
```

Screencaps

image

A few TODOs:
- [x] make sure there is no regression on the frontend from 1.2 to 2.x
- [x] test previews with multiple editors rendering on the same page

## Related Issues

Rebase of #112.

### CODE REVIEW

**Review by ifox — CHANGES_REQUESTED** [review]
Awesome work @ptrckvzn, thanks for submitting! I've added a few comments and will test this out in the next few days.

**Comment by idflood** [comment]
Awesome news : )

Now I'm wondering if these changes would make the whole "blocks" translatable? Currently it's only possible to [translate inner block](https://github.com/area17/twill/issues/204) fields, but for most of my websites the blocks could be different between languages.

I totally understand if it would still not be possible, but since there is a lot of refactoring going on here maybe a small change now would make translatable blocks "easy" to implement in the future.

File: frontend/js/components/Dropdown.vue [hunk 1]
```diff
@@ -107,7 +107,7 @@
         return {
           'max-height': this.maxHeight > 0 ? this.maxHeight + 'px' : '',
           overflow: this.maxHeight > 0 ? 'hidden' : '',
-          'overflow-y': this.maxHeight > 0 ? 'scroll' : ''
+          'overflow-y': this.maxHeight > 0 ? 'auto' : ''
```
[the remark this answers is not in the document]
Comment (ptrckvzn) [reply frontend/js/components/Dropdown.vue]: This isn't necessary anymore I believe, so undoing the change.

**Comment by ifox** [comment]
Hi @agnonym, you're totally right. I updated the mapping [here](https://github.com/area17/twill/commit/397792cafec7a43ec732ea22c9d1d3fa46b8e5f3) before releasing it for backward compatibility. In Twill 3.0 we will do some cleanup and realignment of the form fields options naming, but for now, I had to do this to avoid weird results.

**Comment by ifox** [comment]
I updated @ptrckvzn's description to reflect that change:

```
@formField('block_editor', [
  'name' => 'fantastic-name',
  'title' => 'Fantastic name', // title to display in the Editor dropdown selector
  'label' => 'Add fantastic content', // custom trigger label
  'blocks' => ['text', 'image', 'accordion'],
]);
```

**Pull request merged**